### ISSUE TITLE
x509 support for docker-registry with fine grain control.

### ISSUE THREAD

**Issue by clemenko** [docker/distribution#107]
We have a need for x509 (cert) support for authenticating docker daemons. We can achieve this now with an apache(nginx) proxy for the tls termination. However we can't apply fine grain controls to the namespaces preventing images from being overwritten. We are essentially looking for the same "level" of control with "docker login" but using x509 certs. DHE may ultimately provide the solution for this.

**Comment by dmp42** [comment]
cc @dmcgowan @stevvooe

cc @BrianBland from DHE

**Comment by stevvooe** [comment]
@clemenko If I understand the request, you'd like to have a certificate set the permissions available to a user of the registry. Effectively, the registry auth system would trust tokens and internal scopes signed by the certificate but the available scopes (push/pull by repository) would be restricted the by the certificate.

If this doesn't sound correct, could you lay out a few use cases describing how the docker client, docker daemon and registry would interact? Please be specific about the contents of the certificate and the interaction with docker login and the authorization scopes in the registry.

Please let me know if you have questions.

**Comment by clemenko** [comment]
@stevvooe We would like the DN from the x509 Cert act as the username. And then the regular "docker login" user restrictions still apply. We effectively want to replace the username/password with Certs.
Example : 
userA has a cert with a DN of "userA". The configure the daemon to use that cert when talking to the registry. They can pull from the "library" of public images ( docker pull server/centos). And they can push and pull to the namespace of "userA" (docker pull server/userA/web). UserA can ONLY PUSH to images within their namespace (docker push server/userA/web). 
UserB can pull userA's images but can't PUSH images to that namespace. aka. 
docker pull server/userA/web - works
docker push server/userA/web - FATA[000].

I am trying to achieve this with apache mod_rewrite. But I am hoping to have this built in.
make sense?

**Comment by stevvooe** [comment]
@clemenko Yes, this does make sense. For more background, the registry authentication system is described [here](http://godoc.org/github.com/docker/distribution/auth), along with [token-based auth](http://godoc.org/github.com/docker/distribution/auth/token). 

Where would the permissions be stored? Is that within the x509 certificate or could that be in the server's certificate chain? The namespace model is simple but very limiting. What if we wanted a given user to be able to push to other namespaces? What if the concept of namespaces changed?

cc @jlhawn @dmcgowan

**Comment by clemenko** [comment]
Permissions for users to push to other namespaces is a bit trickier. Currently under the "basic auth" can userA overwrite userB's images? I was hope that DHE would be able to handle that because it introduces groups and larger overrides. Keep in mind that the x509 cert is only replacing the username and password. It is a signed cert that is verified against a Certificate Authority. Just like a user supplied password is verified against a local hash/store.

**Comment by stevvooe** [comment]
@clemenko I can't comment on DHE since I don't know. Please ask @BrianBland @AndreyKostov.

Under "basic auth" (proxied via nginx or apache), the registry authorization is "open" after authentication. However, the internal v2 auth model is not user oriented: is it user + scope oriented. This *could* be supported by x509 authorities without network round trips but there are complexities in mapping scopes to users that need to be worked out.

I think we have enough information about your use case to inform the direction of auth in the registry. We'll take these into account when iterating on these features.

Thank you for clarifying!

**Comment by clemenko** [comment]
So right now under "basic auth" userA can overwrite userB's images, assuming server/userA/image namespaces?

**Comment by stevvooe** [comment]
@clemenko It depends on the configuration of basic auth coming _only into the registry through a proxy_. For a naive username and password checks, yes, userA could overwrite userB images. With a richer proxy configuration, one can restrict modifications based on url and method to prevent this case. The [V2 URL layout](https://github.com/docker/distribution/blob/master/doc/SPEC.md#detail) was designed to allow for these richer use cases implemented in a proxy.

__This only applies to the case where a registry is behind a proxy implementing basic auth.__

**Comment by clemenko** [comment]
@stevvooe Do you know anyone that has already made the "richer" proxy configuration. It would save me some time. It will be nice to see v2, and docker 1.5 rpms released.

**Comment by dmp42** [comment]
@clemenko neither v2 or docker 1.5 are released yet and documentation is sparse :)
Proxies with fine-grain ACLs are a mere possibility for now (though we made sure that was easier to do with v2 than with v1).

Although, feedback and feature requests are always welcome (preferably, as detailed as possible about the use case).

About DHE, I encourage you to reach out to sales (docker.com) as we only handle the open-source components here.

Hope that helps!

**Comment by clemenko** [comment]
I have reached out to DHE and they have no tech specs yet. ETA on v2?
